chapter/related-work.md

```diff
@@ -6,6 +6,90 @@ As such, it is important to get a picture of the field of current LSP projects.
+
+### IDE Abstraction
+
+#### Monto
+
+The authors of the Monto project[@monto,@monto-disintegrated] call this the "IDE Portability Problem".
+They compare the situation with the task of compiling different high level languages to a set of CPU architectures.
+The answer to that problem was an intermediate representation (IR).
+Compilers could transform input languages into this IR and in turn generate assembly for different architectures from a single input format.
+
+With Monto, Kreidel et al propose a similar idea for IDE portability.
+The paper describes the *Monto IR* and how they use a *Message Broker* to receive events from the Editor and dispatch them to *Monto Services*.
+
+The Monto IR is a language agnostic and editor independent, JSON serialized tree-like model.
+Additionally, the IR maintains low level syntax highlighting information (font, color, style, etc.) but leaves the highlighting to the language specific service.
+
+The processing and modification of the source code and IR is performed by *Monto Services*.
+Services implement specific actions, e.g. parsing, outlining or highlighting.
+A central broker connects the services with each other and the editor.
+
+Since Monto performs all work on the IR, independent of the editor, and serializes the IR as JSON messages, the language used to implement *Monto Services* can be chosen freely giving even more flexibility.
+
+The Editor extension's responsibility is to act as a source and sink for data.
+It sends Monto compliant messages to the broker and receives processing results such as (error) reports.
+The communication is based on the ZeroMQ[zeromq] technology which was chosen because it is lightweight and available in manly languages [@monto-disintegrated] allowing to make use of existing language tools.
+
+#### Merlin
+
+The Merlin tool [@merlin,@merlin-website] is in many ways a more specific version of the idea presented in Monto.
+Merlin is a language server for the Ocaml language, yet predates the Language Server Protocol.
+
+The authors of Merlin postulate that implementing "tooling support traditionally provided by IDEs" for "niche languages" demands to "share the language-awareness logic" between implementations.
+As an answer to that, they describe the architecture of Merlin in [@merlin].
+
+Similarly to Monto, Merlin separates editor extensions from language analysis.
+Conversely, its interaction builds on a command line interface instead of message passing.
+Editor extensions expose the server functions to the user by integrating with the editor.
+
+The Merlin server hand provides a single optimized implementation of code intelligence for Ocaml.
+Since all resources could be put to a single project, multiple iterations of performance improvements were done on Merlin.
+It now supports partial, incremental parsing and type-checking which allows users to query information even about incomplete or incorrect programs.
```

Comment:
The 3 lines above are slightly heavy. Suggestion:

"Thanks to this architecture, the Merlin developers have been able to put all their effort in a single server providing code intelligence for Ocaml source code. As a result, Merlin is performant, features fault-tolerant and incremental parsing as well as incremental typechecking, which allows users to query information even about incomplete or incorrect programs.

Reply:
see: https://github.com/ysndr/MSc-Thesis/pull/18/files#r820110050